### PR TITLE
[netcore] Ongoing work.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: mcs/class/System.Private.CoreLib/System.Threading/WaitHandle.cs [hunk 1]
```diff
@@ -4,14 +4,29 @@ namespace System.Threading
...
 		}
 
 		[MethodImpl (MethodImplOptions.InternalCall)]
 		static extern int SignalAndWaitCore (IntPtr waitHandleToSignal, IntPtr waitHandleToWaitOn, int millisecondsTimeout);
 
-		internal static int WaitMultipleIgnoringSyncContext (Span<IntPtr> waitHandles, bool waitAll, int millisecondsTimeout) => throw new PlatformNotSupportedException ();
+		internal static int WaitMultipleIgnoringSyncContext (Span<IntPtr> waitHandles, bool waitAll, int millisecondsTimeout)
+		{
+			unsafe {
+				IntPtr* handles = stackalloc IntPtr[waitHandles.Length];
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Threading/WaitHandle.cs]: Use MemoryMarshal.GetReference instead

File: mcs/class/System.Private.CoreLib/System.Threading/WaitHandle.cs [hunk 2]
```diff
@@ -4,14 +4,29 @@ namespace System.Threading
 {
 	partial class WaitHandle
 	{
+		[MethodImplAttribute(MethodImplOptions.InternalCall)]
+		internal unsafe static extern int Wait_internal(IntPtr* handles, int numHandles, bool waitAll, int ms);
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Threading/WaitHandle.cs]: Remove `internal`

File: mcs/class/System.Private.CoreLib/System.Reflection/RuntimeMethodBody.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	internal sealed class RuntimeMethodBody : MethodBody
+	{
+		ExceptionHandlingClause[] clauses;
+		LocalVariableInfo[] locals;
+		byte[] il;
+		bool init_locals;
+		int sig_token;
+		int max_stack;
+
+		protected RuntimeMethodBody ()
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Reflection/RuntimeMethodBody.cs]: Why is it needed when it's sealed and runtime call the overload bellow?

File: mcs/class/System.Private.CoreLib/System.Reflection/RuntimeLocalVariableInfo.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Reflection {
+	[StructLayout (LayoutKind.Sequential)]
+	public class RuntimeLocalVariableInfo : LocalVariableInfo {
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Reflection/RuntimeLocalVariableInfo.cs]: It should be public

File: mcs/class/System.Private.CoreLib/System.Reflection/RuntimeLocalVariableInfo.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+		#region Keep in sync with MonoReflectionLocalVariableInfo in object-internals.h
+		internal Type type;
+		internal bool is_pinned;
+		internal ushort position;
+		#endregion
+
+		protected RuntimeLocalVariableInfo ()
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Reflection/RuntimeLocalVariableInfo.cs]: Why protected?

File: mcs/class/System.Private.CoreLib/System.Reflection/RuntimeExceptionHandlingClause.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+		internal Type catch_type;
+		internal int filter_offset;
+		internal ExceptionHandlingClauseOptions flags;
+		internal int try_offset;
+		internal int try_length;
+		internal int handler_offset;
+		internal int handler_length;
+		#endregion
+
+		protected RuntimeExceptionHandlingClause ()
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Reflection/RuntimeExceptionHandlingClause.cs]: Why protected when the type is sealed?

File: mcs/class/System.Private.CoreLib/System.Reflection/MethodBase.cs [hunk 7]
```diff
@@ -1,11 +1,38 @@
...
 namespace System.Reflection
 {
 	partial class MethodBase
 	{
-		public static MethodBase GetMethodFromHandle (RuntimeMethodHandle handle) { throw null; }
-		public static MethodBase GetMethodFromHandle (RuntimeMethodHandle handle, RuntimeTypeHandle declaringType) { throw null; }
+		public static MethodBase GetMethodFromHandle (RuntimeMethodHandle handle)
+		{
+			if (handle.IsNullHandle ())
+				throw new ArgumentException (Environment.GetResourceString("Argument_InvalidHandle"));
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Reflection/MethodBase.cs]: Use ` throw new ArgumentException (SR.Argument_InvalidHandle);` instead

File: mcs/class/System.Private.CoreLib/System.Reflection/MethodBase.cs [hunk 8]
```diff
@@ -1,11 +1,38 @@
...
+				throw new ArgumentException (String.Format (Environment.GetResourceString ("Argument_MethodDeclaringTypeGeneric"),
+															m, declaringType.GetGenericTypeDefinition ()));
+
+			return m;
+		}
+
+		public static MethodBase GetMethodFromHandle (RuntimeMethodHandle handle, RuntimeTypeHandle declaringType)
+		{
+			if (handle.IsNullHandle ())
+				throw new ArgumentException (Environment.GetResourceString("Argument_InvalidHandle"));
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Reflection/MethodBase.cs]: Use throw new ArgumentException (SR.Argument_InvalidHandle); instead

File: mcs/class/System.Private.CoreLib/System.Reflection/MethodBase.cs [hunk 9]
```diff
@@ -1,11 +1,38 @@
...
+
+			MethodBase m = RuntimeMethodInfo.GetMethodFromHandleInternalType (handle.Value, IntPtr.Zero);
+			if (m == null)
+				throw new ArgumentException ("The handle is invalid.");
+
+			Type declaringType = m.DeclaringType;
+			if (declaringType != null && declaringType.IsGenericType)
+				throw new ArgumentException (String.Format (Environment.GetResourceString ("Argument_MethodDeclaringTypeGeneric"),
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Reflection/MethodBase.cs]: Use `SR.Argument_MethodDeclaringTypeGeneric` constant instead

File: mcs/class/System.Private.CoreLib/System.Reflection/FieldInfo.cs [hunk 10]
```diff
@@ -1,8 +1,105 @@
...
 	partial class FieldInfo
 	{
-		public static FieldInfo GetFieldFromHandle (RuntimeFieldHandle handle) { throw null; }
-		public static FieldInfo GetFieldFromHandle (RuntimeFieldHandle handle, RuntimeTypeHandle declaringType) { throw null; }		
+		[MethodImplAttribute(MethodImplOptions.InternalCall)]
+		private static extern FieldInfo internal_from_handle_type (IntPtr field_handle, IntPtr type_handle);
+
+		public static FieldInfo GetFieldFromHandle (RuntimeFieldHandle handle)
+		{
+			if (handle.Value == IntPtr.Zero)
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Reflection/FieldInfo.cs]: Would be better to use
```
if (handle.IsNullHandle ())
	throw new ArgumentException (SR. Argument_InvalidHandle);
```

for consistency

File: mcs/class/System.Private.CoreLib/System.Reflection/FieldInfo.cs [hunk 11]
```diff
@@ -1,8 +1,105 @@
...
+				throw new ArgumentException ("The handle is invalid.");
+			return internal_from_handle_type (handle.Value, IntPtr.Zero);
+		}
+
+		public static FieldInfo GetFieldFromHandle (RuntimeFieldHandle handle, RuntimeTypeHandle declaringType)
+		{
+			if (handle.Value == IntPtr.Zero)
```
Comment (marek-safar) [inline mcs/class/System.Private.CoreLib/System.Reflection/FieldInfo.cs]: Same as above

**Comment by marek-safar** [comment]
@vargax breaks existing tests (* Assertion at object.c:5317, condition `arg->vtable->klass == mono_defaults.int_class' not met)

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@monojenkins build

**Pull request merged**